common/flatpak-dir.c

```diff
@@ -7982,6 +8032,80 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
   return NULL;
 }
 
+static GPtrArray *
+filter_out_deployed_refs (FlatpakDir *self,
+                          GPtrArray  *local_refspecs,
+                          GError     **error)
+{
+  g_autoptr(GPtrArray) undeployed_refs = g_ptr_array_new_full (local_refspecs->len, g_free);
+  g_autoptr(GVariant) deploy_data = NULL;
+  gsize i = 0;
+
+  for (; i < local_refspecs->len; ++i)
+    {
+      const gchar *refspec = g_ptr_array_index (local_refspecs, i);
+      g_autofree gchar *remote = NULL;
+      g_autofree gchar *ref = NULL;
+
+      if (!ostree_parse_refspec (refspec, &remote, &ref, error))
+        return FALSE;
+
+      g_autoptr(GVariant) deploy_data = flatpak_dir_get_deploy_data (self,
+                                                                     ref,
+                                                                     NULL,
+                                                                     NULL);
+
+      if (!deploy_data)
+        g_ptr_array_add (undeployed_refs, g_strdup (refspec));
+    }
+
+  return g_steal_pointer (&undeployed_refs);
+}
+
+gboolean
+flatpak_dir_cleanup_undeployed_refs (FlatpakDir   *self,
```

Comment:
Nitpick: It would be good to have a comment above this function summarising what it does; makes the code a bit easier to understand at a glance.